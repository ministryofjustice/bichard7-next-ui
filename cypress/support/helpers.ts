import CaseDetailsTab from "types/CaseDetailsTab"

export function confirmFiltersAppliedContains(filterTag: string) {
  cy.get(".moj-filter-tags a.moj-filter__tag").contains(filterTag)
}

export const exactMatch = (keyword: string): RegExp => {
  return new RegExp("^" + keyword + "$")
}

export const confirmMultipleFieldsDisplayed = (fields: string[]) => {
  fields.forEach((field) => {
    cy.contains(field)
  })
}

export const confirmMultipleFieldsNotDisplayed = (fields: string[]) => {
  fields.forEach((field) => {
    cy.contains(field).should("not.exist")
  })
}

export const removeFilterChip = () => {
  cy.get("li button.moj-filter__tag").trigger("click")
  cy.get(".moj-filter__tag").should("not.exist")
}

export const filterByCaseAge = (caseAgeId: string) => {
  cy.get("#case-age").click()
  cy.get(caseAgeId).click()
}

export const filterByDateRange = (dateFrom: string, dateTo: string) => {
  cy.get("#date-range").click()

  cy.get("#date-from").click()
  cy.get("#date-from").type(dateFrom)

  cy.get("#date-to").click()
  cy.get("#date-to").type(dateTo)
}

export const loginAndGoToUrl = (emailAddress = "bichard01@example.com", url = "/bichard") => {
  cy.login(emailAddress, "password")
  cy.visit(url)
}

export const clickTab = (tab: CaseDetailsTab) => {
  cy.contains(tab).click()
  cy.get("H3").contains(tab)
}
