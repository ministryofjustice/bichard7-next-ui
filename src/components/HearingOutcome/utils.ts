import { AmendmentArrValues, AmendmentValues, IndividualAmendmentArrValues } from "types/Amendments"

const isAmendmentArrValue = (value: AmendmentValues, indexToUpdate: number): value is AmendmentArrValues =>
  (value as AmendmentArrValues)[indexToUpdate].updatedValue !== undefined

export const isAmendmentArr = (value: AmendmentValues): value is AmendmentArrValues => typeof value !== "string"

export const isAmendmentValue = (value: unknown): value is AmendmentValues => {
  if (typeof value === "string") {
    return true
  }

  // if it's an array with these specific keys on then it's also an AmendmentValues
  return Array.isArray(value) ? value.every((v) => "offenceIndex" in v && "updatedValue" in v) : false
}

// check to see if an object for this value already exists
export const doesUpdateExist = (
  amendmentsArr: AmendmentArrValues,
  incomingValue: IndividualAmendmentArrValues
): number =>
  amendmentsArr.findIndex((existingUpdate: IndividualAmendmentArrValues) => {
    const correctOffenceIdx = existingUpdate.offenceIndex === incomingValue.offenceIndex

    if ("resultIndex" in incomingValue && correctOffenceIdx) {
      if ("resultIndex" in existingUpdate) {
        if (existingUpdate.resultIndex === incomingValue.resultIndex) {
          return correctOffenceIdx
        } else {
          return -1
        }
      }
    }

    return correctOffenceIdx
  })

export const appendNewValue = (
  newArrValue: IndividualAmendmentArrValues,
  keyToAmend: string,
  indexToUpdate: number,
  amendmentsObj: Record<string, AmendmentValues>
) => {
  if (indexToUpdate > -1 && isAmendmentArrValue(amendmentsObj[keyToAmend], indexToUpdate)) {
    // updated the existing object
    ;(amendmentsObj[keyToAmend] as AmendmentArrValues)[indexToUpdate].updatedValue = newArrValue.updatedValue
    return [...amendmentsObj[keyToAmend].slice()]
  }

  // it doesn't already exist so append it
  return [...amendmentsObj[keyToAmend].slice(), newArrValue]
}
