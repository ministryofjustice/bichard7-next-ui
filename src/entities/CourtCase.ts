/* eslint-disable import/no-cycle */
import { Column, Entity, OneToMany, PrimaryColumn } from "typeorm"
import type { Relation } from "typeorm"
import BaseEntity from "./BaseEntity"
import dateTransformer from "./transformers/dateTransformer"
import Note from "./Note"
import Trigger from "./Trigger"
import ahoTransformer from "./transformers/ahoTransformer"
import type { AnnotatedHearingOutcome } from "@moj-bichard7-developers/bichard7-next-core/build/src/types/AnnotatedHearingOutcome"

@Entity({ name: "error_list" })
export default class CourtCase extends BaseEntity {
  @PrimaryColumn({ name: "error_id" })
  errorId!: number

  @Column({ name: "message_id" })
  messageId!: string

  @Column({ name: "court_name" })
  courtName!: string

  @Column({ name: "court_date", type: "date", nullable: true, transformer: dateTransformer })
  courtDate?: Date

  @Column({ name: "ptiurn" })
  ptiurn!: string

  @Column({ name: "defendant_name" })
  defendantName!: string

  @Column({ name: "error_reason" })
  errorReason!: string

  @Column({ name: "trigger_reason" })
  triggerReason!: string

  @Column({ name: "org_for_police_filter" })
  orgForPoliceFilter!: string

  @OneToMany(() => Trigger, (trigger) => trigger.courtCase)
  triggers!: Relation<Trigger>[]

  @Column({ name: "trigger_count" })
  triggerCount!: number

  @Column({ name: "error_count" })
  errorCount!: number

  @Column({ name: "annotated_msg", type: "varchar", transformer: ahoTransformer })
  ahoXml!: AnnotatedHearingOutcome

  @OneToMany(() => Note, (note) => note.courtCase)
  notes!: Relation<Note>[]
}
