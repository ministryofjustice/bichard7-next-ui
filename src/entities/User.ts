import { Column, Entity, PrimaryColumn } from "typeorm"
import BaseEntity from "./BaseEntity"
import delimeterStringTransfomer from "./transformers/delimeterStringTransfomer"

@Entity({ name: "users" })
export default class User extends BaseEntity {
  @PrimaryColumn()
  username!: string

  @Column()
  forenames?: string

  @Column()
  surname?: string

  @Column({ name: "visible_forces", transformer: delimeterStringTransfomer(","), type: "varchar" })
  visibleForces!: string[]
}
