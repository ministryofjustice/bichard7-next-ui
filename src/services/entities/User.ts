import { Column, Entity, PrimaryColumn } from "typeorm"
import GroupName from "types/GroupName"
import type KeyValuePair from "types/KeyValuePair"
import BaseEntity from "./BaseEntity"
import delimitedPrefixedString from "./transformers/delimitedPrefixedString"
import featureFlagTransformer from "./transformers/featureFlagTransformer"

@Entity({ name: "users" })
export default class User extends BaseEntity {
  @PrimaryColumn()
  username!: string

  @Column()
  email!: string

  @Column()
  forenames?: string

  @Column()
  surname?: string

  @Column({ name: "visible_forces", transformer: delimitedPrefixedString(",", "0"), type: "varchar" })
  visibleForces!: string[]

  @Column({ name: "feature_flags", transformer: featureFlagTransformer, type: "jsonb" })
  featureFlags!: KeyValuePair<string, boolean>

  private _groups: GroupName[] = []

  get groups() {
    return this._groups
  }

  set groups(value: GroupName[]) {
    this._groups = value
  }

  get canLockTriggers() {
    return this._groups.some(
      (group) =>
        group === "TriggerHandler" || group === "GeneralHandler" || group === "Allocator" || group === "Supervisor"
    )
  }

  get canLockExceptions() {
    return this._groups.some(
      (group) =>
        group === "ExceptionHandler" || group === "GeneralHandler" || group === "Allocator" || group === "Supervisor"
    )
  }

  get canUnlockOtherUsersTriggers() {
    return this._groups.some((group) => group === "Allocator" || group === "Supervisor")
  }

  get canUnlockOtherUsersExceptions() {
    return this._groups.some((group) => group === "Allocator" || group === "Supervisor")
  }
}
