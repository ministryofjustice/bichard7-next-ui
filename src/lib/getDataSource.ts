import CourtCase from "../entities/CourtCase"
import { DataSource } from "typeorm"
import DatabaseConfig from "./DatabaseConfig"

const databaseConfig: DatabaseConfig = {
  host: process.env.DB_HOST ?? process.env.DB_AUTH_HOST ?? "localhost",
  user: process.env.DB_USER ?? process.env.DB_AUTH_USER ?? "bichard",
  password: process.env.DB_PASSWORD ?? process.env.DB_AUTH_PASSWORD ?? "password",
  database: process.env.DB_DATABASE ?? process.env.DB_AUTH_DATABASE ?? "bichard",
  port: parseInt(process.env.DB_PORT ?? process.env.DB_AUTH_PORT ?? "5432", 10),
  ssl: (process.env.DB_SSL ?? process.env.DB_AUTH_SSL) === "true",
  schema: "br7own"
}

const getDataSource = async (): Promise<DataSource> => {
  const appDataSource = new DataSource({
    type: "postgres",
    applicationName: "ui-connection",
    host: databaseConfig.host,
    port: databaseConfig.port,
    username: databaseConfig.user,
    password: databaseConfig.password,
    database: databaseConfig.database,
    synchronize: false,
    logging: false,
    entities: [CourtCase],
    subscribers: [],
    migrations: [],
    schema: databaseConfig.schema,
    ssl: databaseConfig.ssl ? { rejectUnauthorized: false } : false,
    extra: {
      max: 1
    }
  })

  if (!appDataSource.isInitialized) {
    await appDataSource.initialize()
  }

  return appDataSource
}

export default getDataSource
