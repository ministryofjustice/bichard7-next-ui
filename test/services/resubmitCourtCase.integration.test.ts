import parseAhoXml from "@moj-bichard7-developers/bichard7-next-core/dist/parse/parseAhoXml/parseAhoXml"
import { AnnotatedHearingOutcome } from "@moj-bichard7-developers/bichard7-next-core/dist/types/AnnotatedHearingOutcome"
import CourtCase from "services/entities/CourtCase"
import User from "services/entities/User"
import getDataSource from "services/getDataSource"
import insertNotes from "services/insertNotes"
import sendToQueue from "services/mq/sendToQueue"
import resubmitCourtCase from "services/resubmitCourtCase"
import { DataSource } from "typeorm"
import offenceSequenceException from "../test-data/HO100302_1.json"
import deleteFromEntity from "../utils/deleteFromEntity"
import { getDummyCourtCase, insertCourtCases } from "../utils/insertCourtCases"
import { hasAccessToAll } from "../helpers/hasAccessTo"

jest.mock("services/mq/sendToQueue")
jest.mock("services/insertNotes")

jest.setTimeout(60 * 60 * 1000)

describe("resubmit court case", () => {
  const userName = "UserName"
  let dataSource: DataSource

  beforeAll(async () => {
    dataSource = await getDataSource()
  })

  beforeEach(async () => {
    await deleteFromEntity(CourtCase)
    jest.resetAllMocks()
    jest.clearAllMocks()
  })

  afterAll(async () => {
    await dataSource.destroy()
  })

  it("Should resubmit a court case with no updates", async () => {
    // set up court case in the right format to insert into the db
    const inputCourtCase = await getDummyCourtCase({
      errorLockedByUsername: userName,
      triggerLockedByUsername: userName,
      errorCount: 1,
      errorStatus: "Unresolved",
      triggerCount: 1,
      phase: 1,
      hearingOutcome: offenceSequenceException.hearingOutcomeXml,
      updatedHearingOutcome: offenceSequenceException.updatedHearingOutcomeXml,
      orgForPoliceFilter: "1"
    })

    // insert the record to the db
    await insertCourtCases(inputCourtCase)

    // check the queue hasn't been called
    expect(sendToQueue).toHaveBeenCalledTimes(0)

    const result = await resubmitCourtCase(dataSource, { noUpdatesResubmit: true }, inputCourtCase.errorId, {
      username: userName,
      visibleForces: ["1"],
      visibleCourts: [],
      hasAccessTo: hasAccessToAll
    } as Partial<User> as User)

    expect(result).not.toBeInstanceOf(Error)
    expect(result).toMatchSnapshot()

    // pull out the case from the db
    const retrievedCase = await dataSource
      .getRepository(CourtCase)
      .findOne({ where: { errorId: inputCourtCase.errorId } })

    expect(sendToQueue).toHaveBeenCalledTimes(1)
    expect(insertNotes).toHaveBeenCalledTimes(2)
    expect(insertNotes).toHaveBeenCalledWith(expect.anything(), [
      { errorId: inputCourtCase.errorId, noteText: "UserName: Portal Action: Resubmitted Message.", userId: "System" }
    ])

    // assert that the xml in the db is as we expect
    expect(retrievedCase?.updatedHearingOutcome).toMatchSnapshot()
    expect(retrievedCase?.hearingOutcome).toMatchSnapshot()
    expect(retrievedCase?.errorStatus).toBe("Submitted")
  })

  it("Should resubmit a court case with updates to Court Offence Sequence Number", async () => {
    // set up court case in the right format to insert into the db
    const inputCourtCase = await getDummyCourtCase({
      errorLockedByUsername: userName,
      triggerLockedByUsername: null,
      errorCount: 1,
      errorStatus: "Unresolved",
      triggerCount: 1,
      phase: 1,
      hearingOutcome: offenceSequenceException.hearingOutcomeXml,
      updatedHearingOutcome: offenceSequenceException.updatedHearingOutcomeXml,
      orgForPoliceFilter: "1111"
    })

    // insert the record to the db
    await insertCourtCases(inputCourtCase)

    // check the queue hasn't been called
    expect(sendToQueue).toHaveBeenCalledTimes(0)

    // parse the xml so we can assert on the values before they change
    const input = parseAhoXml(inputCourtCase.updatedHearingOutcome as string)

    expect(input).not.toBeInstanceOf(Error)

    expect(
      (input as AnnotatedHearingOutcome).AnnotatedHearingOutcome.HearingOutcome.Case.HearingDefendant.Offence[0]
        .CourtOffenceSequenceNumber
    ).not.toEqual(1234)

    const result = await resubmitCourtCase(
      dataSource,
      { courtOffenceSequenceNumber: [{ offenceIndex: 0, updatedValue: 1234 }] },
      inputCourtCase.errorId,
      {
        username: userName,
        visibleForces: ["1111"],
        visibleCourts: [],
        hasAccessTo: hasAccessToAll
      } as Partial<User> as User
    )

    expect(result).not.toBeInstanceOf(Error)
    expect(result).toMatchSnapshot()

    // pull out the case from the db
    const retrievedCase = await dataSource
      .getRepository(CourtCase)
      .findOne({ where: { errorId: inputCourtCase.errorId } })

    expect(sendToQueue).toHaveBeenCalledTimes(1)
    expect(insertNotes).toHaveBeenCalledTimes(2)
    expect(insertNotes).toHaveBeenCalledWith(expect.anything(), [
      { errorId: inputCourtCase.errorId, noteText: "UserName: Portal Action: Resubmitted Message.", userId: "System" }
    ])

    // parse the retreived case to aho format so we can assert on the values
    const parsedCase = parseAhoXml((retrievedCase as CourtCase).updatedHearingOutcome as string)

    expect(parsedCase).not.toBeInstanceOf(Error)

    expect(
      (parsedCase as AnnotatedHearingOutcome).AnnotatedHearingOutcome.HearingOutcome.Case.HearingDefendant.Offence[0]
        .CourtOffenceSequenceNumber
    ).toEqual(1234)

    // assert that the xml in the db is as we expect
    expect(retrievedCase?.updatedHearingOutcome).toMatchSnapshot()
    expect(retrievedCase?.hearingOutcome).toMatchSnapshot()
    expect(retrievedCase?.errorStatus).toBe("Submitted")
  })

  it("Should resubmit a court case with updates to multiple offences", async () => {
    const amendments = [
      { offenceIndex: 0, updatedValue: 1234 },
      { offenceIndex: 1, updatedValue: 1234 }
    ]

    // set up court case in the right format to insert into the db
    const inputCourtCase = await getDummyCourtCase({
      errorLockedByUsername: userName,
      triggerLockedByUsername: null,
      errorCount: 1,
      errorStatus: "Unresolved",
      triggerCount: 1,
      phase: 1,
      hearingOutcome: offenceSequenceException.hearingOutcomeXml,
      updatedHearingOutcome: offenceSequenceException.updatedHearingOutcomeXml,
      orgForPoliceFilter: "1111"
    })

    // insert the record to the db
    await insertCourtCases(inputCourtCase)

    // check the queue hasn't been called
    expect(sendToQueue).toHaveBeenCalledTimes(0)

    // parse the xml so we can assert on the values before they change
    const input = parseAhoXml(inputCourtCase.updatedHearingOutcome as string)

    expect(input).not.toBeInstanceOf(Error)

    amendments.forEach(({ offenceIndex, updatedValue }) => {
      expect(
        (input as AnnotatedHearingOutcome).AnnotatedHearingOutcome.HearingOutcome.Case.HearingDefendant.Offence[
          offenceIndex
        ].CourtOffenceSequenceNumber
      ).not.toEqual(updatedValue)
    })

    const result = await resubmitCourtCase(
      dataSource,
      {
        courtOffenceSequenceNumber: amendments
      },
      inputCourtCase.errorId,
      {
        username: userName,
        visibleForces: ["1111"],
        visibleCourts: [],
        hasAccessTo: hasAccessToAll
      } as Partial<User> as User
    )

    expect(result).not.toBeInstanceOf(Error)
    expect(result).toMatchSnapshot()

    // pull out the case from the db
    const retrievedCase = await dataSource
      .getRepository(CourtCase)
      .findOne({ where: { errorId: inputCourtCase.errorId } })

    expect(sendToQueue).toHaveBeenCalledTimes(1)
    expect(insertNotes).toHaveBeenCalledTimes(2)
    expect(insertNotes).toHaveBeenCalledWith(expect.anything(), [
      { errorId: inputCourtCase.errorId, noteText: "UserName: Portal Action: Resubmitted Message.", userId: "System" }
    ])

    // parse the retreived case to aho format so we can assert on the values
    const parsedCase = parseAhoXml((retrievedCase as CourtCase).updatedHearingOutcome as string)

    expect(parsedCase).not.toBeInstanceOf(Error)

    amendments.forEach(({ offenceIndex, updatedValue }) => {
      expect(
        (parsedCase as AnnotatedHearingOutcome).AnnotatedHearingOutcome.HearingOutcome.Case.HearingDefendant.Offence[
          offenceIndex
        ].CourtOffenceSequenceNumber
      ).toEqual(updatedValue)
      expect(
        (parsedCase as AnnotatedHearingOutcome).AnnotatedHearingOutcome.HearingOutcome.Case.HearingDefendant.Offence[
          offenceIndex
        ].CourtOffenceSequenceNumber
      ).toEqual(updatedValue)
    })

    // assert that the xml in the db is as we expect
    expect(retrievedCase?.updatedHearingOutcome).toMatchSnapshot()
    expect(retrievedCase?.hearingOutcome).toMatchSnapshot()
    expect(retrievedCase?.errorStatus).toBe("Submitted")
  })
})
