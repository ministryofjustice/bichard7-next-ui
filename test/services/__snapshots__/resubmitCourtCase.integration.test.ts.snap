// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resubmit court case should resubmit a court case with no updates 1`] = `
{
  "AnnotatedHearingOutcome": {
    "HearingOutcome": {
      "Case": {
        "CourtCaseReferenceNumber": undefined,
        "CourtReference": {
          "MagistratesCourtReference": "01ZD0300208",
        },
        "ForceOwner": {
          "BottomLevelCode": "00",
          "OrganisationUnitCode": "01ZD00",
          "SecondLevelCode": "01",
          "ThirdLevelCode": "ZD",
          "TopLevelCode": undefined,
        },
        "HearingDefendant": {
          "Address": {
            "AddressLine1": "ScenarioAE Address Line 1",
            "AddressLine2": "ScenarioAE Address Line 2",
            "AddressLine3": "ScenarioAE Address Line 3",
            "AddressLine4": undefined,
            "AddressLine5": undefined,
          },
          "ArrestSummonsNumber": "1101ZD0100000411423H",
          "BailConditions": [],
          "CourtPNCIdentifier": undefined,
          "DefendantDetail": {
            "BirthDate": 1948-11-11T00:00:00.000Z,
            "Gender": 1,
            "GeneratedPNCFilename": "PINKERTON/MARCUS",
            "PersonName": {
              "FamilyName": "Pinkerton",
              "GivenName": [
                "Marcus",
              ],
              "Title": "Mr",
            },
          },
          "Offence": [
            {
              "ActualOffenceDateCode": "1",
              "ActualOffenceEndDate": {
                "EndDate": 2010-09-23T00:00:00.000Z,
              },
              "ActualOffenceStartDate": {
                "StartDate": 2010-09-21T00:00:00.000Z,
              },
              "ActualOffenceWording": "Theft from shop - shoplifting.",
              "AddedByTheCourt": undefined,
              "AlcoholLevel": undefined,
              "ArrestDate": 2010-09-25T00:00:00.000Z,
              "ChargeDate": 2010-09-25T00:00:00.000Z,
              "CommittedOnBail": "D",
              "ConvictionDate": 2011-10-01T00:00:00.000Z,
              "CourtCaseReferenceNumber": undefined,
              "CourtOffenceSequenceNumber": 1,
              "CriminalProsecutionReference": {
                "DefendantOrOffender": {
                  "CheckDigit": "H",
                  "DefendantOrOffenderSequenceNumber": "00000411423",
                  "OrganisationUnitIdentifierCode": {
                    "BottomLevelCode": "01",
                    "OrganisationUnitCode": "01ZD01",
                    "SecondLevelCode": "01",
                    "ThirdLevelCode": "ZD",
                    "TopLevelCode": undefined,
                  },
                  "Year": "11",
                },
                "OffenceReason": {
                  "OffenceCode": {
                    "ActOrSource": "TH",
                    "FullCode": "TH68010",
                    "Qualifier": undefined,
                    "Reason": "010",
                    "Year": "68",
                    "__type": "NonMatchingOffenceCode",
                  },
                  "__type": "NationalOffenceReason",
                },
                "OffenceReasonSequence": undefined,
              },
              "HomeOfficeClassification": "046/00",
              "LocationOfOffence": "1 Kingston High Street",
              "ManualCourtCaseReference": undefined,
              "ManualSequenceNumber": undefined,
              "NotifiableToHOindicator": true,
              "OffenceCategory": "CE",
              "OffenceTitle": "Theft from a shop",
              "RecordableOnPNCindicator": true,
              "Result": [
                {
                  "AmountSpecifiedInResult": undefined,
                  "BailCondition": [],
                  "CJSresultCode": 1002,
                  "ConvictingCourt": undefined,
                  "CourtType": "MCA",
                  "DateSpecifiedInResult": undefined,
                  "Duration": [
                    {
                      "DurationLength": 12,
                      "DurationType": "Duration",
                      "DurationUnit": "M",
                    },
                  ],
                  "ModeOfTrialReason": "SUM",
                  "NextCourtType": undefined,
                  "NextHearingDate": undefined,
                  "NextHearingTime": undefined,
                  "NextResultSourceOrganisation": undefined,
                  "NumberOfOffencesTIC": undefined,
                  "NumberSpecifiedInResult": undefined,
                  "OffenceRemandStatus": undefined,
                  "PNCAdjudicationExists": undefined,
                  "PNCDisposalType": 1002,
                  "PleaStatus": "NG",
                  "ReasonForOffenceBailConditions": undefined,
                  "ResultApplicableQualifierCode": [],
                  "ResultClass": "Judgement with final result",
                  "ResultHalfLifeHours": 72,
                  "ResultHearingDate": 2011-10-01T00:00:00.000Z,
                  "ResultHearingType": "OTHER",
                  "ResultQualifierVariable": [],
                  "ResultVariableText": "Imprisonment for 12 Months",
                  "SourceOrganisation": {
                    "BottomLevelCode": "01",
                    "OrganisationUnitCode": "B01EF01",
                    "SecondLevelCode": "01",
                    "ThirdLevelCode": "EF",
                    "TopLevelCode": "B",
                  },
                  "Urgent": undefined,
                  "Verdict": "G",
                  "WarrantIssueDate": undefined,
                },
                {
                  "AmountSpecifiedInResult": [
                    {
                      "Amount": 100,
                      "DecimalPlaces": 2,
                      "Type": "Fine",
                    },
                  ],
                  "BailCondition": [],
                  "CJSresultCode": 1015,
                  "ConvictingCourt": undefined,
                  "CourtType": "MCA",
                  "DateSpecifiedInResult": undefined,
                  "Duration": [],
                  "ModeOfTrialReason": "SUM",
                  "NextCourtType": undefined,
                  "NextHearingDate": undefined,
                  "NextHearingTime": undefined,
                  "NextResultSourceOrganisation": undefined,
                  "NumberOfOffencesTIC": undefined,
                  "NumberSpecifiedInResult": undefined,
                  "OffenceRemandStatus": undefined,
                  "PNCAdjudicationExists": undefined,
                  "PNCDisposalType": 1015,
                  "PleaStatus": "NG",
                  "ReasonForOffenceBailConditions": undefined,
                  "ResultApplicableQualifierCode": [],
                  "ResultClass": "Judgement with final result",
                  "ResultHalfLifeHours": 72,
                  "ResultHearingDate": 2011-10-01T00:00:00.000Z,
                  "ResultHearingType": "OTHER",
                  "ResultQualifierVariable": [],
                  "ResultVariableText": "Fined 100.",
                  "SourceOrganisation": {
                    "BottomLevelCode": "01",
                    "OrganisationUnitCode": "B01EF01",
                    "SecondLevelCode": "01",
                    "ThirdLevelCode": "EF",
                    "TopLevelCode": "B",
                  },
                  "Urgent": undefined,
                  "Verdict": "G",
                  "WarrantIssueDate": undefined,
                },
              ],
            },
            {
              "ActualOffenceDateCode": "1",
              "ActualOffenceEndDate": {
                "EndDate": 2010-09-23T00:00:00.000Z,
              },
              "ActualOffenceStartDate": {
                "StartDate": 2010-09-21T00:00:00.000Z,
              },
              "ActualOffenceWording": "Theft from shop - shoplifting.",
              "AddedByTheCourt": undefined,
              "AlcoholLevel": undefined,
              "ArrestDate": 2010-09-25T00:00:00.000Z,
              "ChargeDate": 2010-09-25T00:00:00.000Z,
              "CommittedOnBail": "D",
              "ConvictionDate": 2011-10-01T00:00:00.000Z,
              "CourtCaseReferenceNumber": undefined,
              "CourtOffenceSequenceNumber": 2,
              "CriminalProsecutionReference": {
                "DefendantOrOffender": {
                  "CheckDigit": "H",
                  "DefendantOrOffenderSequenceNumber": "00000411423",
                  "OrganisationUnitIdentifierCode": {
                    "BottomLevelCode": "01",
                    "OrganisationUnitCode": "01ZD01",
                    "SecondLevelCode": "01",
                    "ThirdLevelCode": "ZD",
                    "TopLevelCode": undefined,
                  },
                  "Year": "11",
                },
                "OffenceReason": {
                  "OffenceCode": {
                    "ActOrSource": "TH",
                    "FullCode": "TH68010",
                    "Qualifier": undefined,
                    "Reason": "010",
                    "Year": "68",
                    "__type": "NonMatchingOffenceCode",
                  },
                  "__type": "NationalOffenceReason",
                },
                "OffenceReasonSequence": undefined,
              },
              "HomeOfficeClassification": "046/00",
              "LocationOfOffence": "1 Kingston High Street",
              "ManualCourtCaseReference": undefined,
              "ManualSequenceNumber": undefined,
              "NotifiableToHOindicator": true,
              "OffenceCategory": "CE",
              "OffenceTitle": "Theft from a shop",
              "RecordableOnPNCindicator": true,
              "Result": [
                {
                  "AmountSpecifiedInResult": [
                    {
                      "Amount": 300,
                      "DecimalPlaces": 2,
                      "Type": "Fine",
                    },
                  ],
                  "BailCondition": [],
                  "CJSresultCode": 1015,
                  "ConvictingCourt": undefined,
                  "CourtType": "MCA",
                  "DateSpecifiedInResult": undefined,
                  "Duration": [],
                  "ModeOfTrialReason": "SUM",
                  "NextCourtType": undefined,
                  "NextHearingDate": undefined,
                  "NextHearingTime": undefined,
                  "NextResultSourceOrganisation": undefined,
                  "NumberOfOffencesTIC": undefined,
                  "NumberSpecifiedInResult": undefined,
                  "OffenceRemandStatus": undefined,
                  "PNCAdjudicationExists": undefined,
                  "PNCDisposalType": 1015,
                  "PleaStatus": "NG",
                  "ReasonForOffenceBailConditions": undefined,
                  "ResultApplicableQualifierCode": [],
                  "ResultClass": "Judgement with final result",
                  "ResultHalfLifeHours": 72,
                  "ResultHearingDate": 2011-10-01T00:00:00.000Z,
                  "ResultHearingType": "OTHER",
                  "ResultQualifierVariable": [],
                  "ResultVariableText": "Fined 300.",
                  "SourceOrganisation": {
                    "BottomLevelCode": "01",
                    "OrganisationUnitCode": "B01EF01",
                    "SecondLevelCode": "01",
                    "ThirdLevelCode": "EF",
                    "TopLevelCode": "B",
                  },
                  "Urgent": undefined,
                  "Verdict": "G",
                  "WarrantIssueDate": undefined,
                },
                {
                  "AmountSpecifiedInResult": undefined,
                  "BailCondition": [],
                  "CJSresultCode": 1002,
                  "ConvictingCourt": undefined,
                  "CourtType": "MCA",
                  "DateSpecifiedInResult": undefined,
                  "Duration": [
                    {
                      "DurationLength": 12,
                      "DurationType": "Duration",
                      "DurationUnit": "M",
                    },
                  ],
                  "ModeOfTrialReason": "SUM",
                  "NextCourtType": undefined,
                  "NextHearingDate": undefined,
                  "NextHearingTime": undefined,
                  "NextResultSourceOrganisation": undefined,
                  "NumberOfOffencesTIC": undefined,
                  "NumberSpecifiedInResult": undefined,
                  "OffenceRemandStatus": undefined,
                  "PNCAdjudicationExists": undefined,
                  "PNCDisposalType": 1002,
                  "PleaStatus": "NG",
                  "ReasonForOffenceBailConditions": undefined,
                  "ResultApplicableQualifierCode": [],
                  "ResultClass": "Judgement with final result",
                  "ResultHalfLifeHours": 72,
                  "ResultHearingDate": 2011-10-01T00:00:00.000Z,
                  "ResultHearingType": "OTHER",
                  "ResultQualifierVariable": [],
                  "ResultVariableText": "Imprisonment for 12 Months",
                  "SourceOrganisation": {
                    "BottomLevelCode": "01",
                    "OrganisationUnitCode": "B01EF01",
                    "SecondLevelCode": "01",
                    "ThirdLevelCode": "EF",
                    "TopLevelCode": "B",
                  },
                  "Urgent": undefined,
                  "Verdict": "G",
                  "WarrantIssueDate": undefined,
                },
              ],
            },
          ],
          "OrganisationName": undefined,
          "PNCCheckname": undefined,
          "PNCIdentifier": undefined,
          "ReasonForBailConditions": undefined,
          "RemandStatus": "UB",
        },
        "ManualForceOwner": undefined,
        "PTIURN": "01ZD0300208",
        "PenaltyNoticeCaseReferenceNumber": undefined,
        "PreChargeDecisionIndicator": false,
        "RecordableOnPNCindicator": true,
        "Urgent": undefined,
      },
      "Hearing": {
        "CourtHearingLocation": {
          "BottomLevelCode": "01",
          "OrganisationUnitCode": "B01EF01",
          "SecondLevelCode": "01",
          "ThirdLevelCode": "EF",
          "TopLevelCode": "B",
        },
        "CourtHouseCode": 2576,
        "CourtHouseName": "Magistrates' Courts London Croydon",
        "CourtType": "MCA",
        "DateOfHearing": 2011-10-01T00:00:00.000Z,
        "DefendantPresentAtHearing": "A",
        "HearingDocumentationLanguage": "D",
        "HearingLanguage": "D",
        "SourceReference": {
          "DocumentName": "SPI Marcus Pinkerton",
          "DocumentType": "SPI Case Result",
          "UniqueID": "CID-a7428cb6-b0d7-4921-a2cf-3b767d4f4c49",
        },
        "TimeOfHearing": "10:00",
      },
    },
  },
  "Exceptions": [],
  "PncErrorMessage": undefined,
  "PncQuery": undefined,
  "PncQueryDate": undefined,
}
`;
